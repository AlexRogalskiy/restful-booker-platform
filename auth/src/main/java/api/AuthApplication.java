package api;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.ComponentScan;

@SpringBootApplication
@ComponentScan()
public class AuthApplication {

    // Testing the lookup by making a change in search
    public static void main(String[] args) {
        SpringApplication.run(AuthApplication.class, args);
    }

}
