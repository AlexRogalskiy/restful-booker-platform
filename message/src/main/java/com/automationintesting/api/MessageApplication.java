package com.automationintesting.api;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.ComponentScan;

@SpringBootApplication
@ComponentScan()
public class MessageApplication {

    public static void main(String[] args) {
        SpringApplication.run(MessageApplication.class, args);
    }

}
