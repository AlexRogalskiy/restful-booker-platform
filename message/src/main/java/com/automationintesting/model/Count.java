package com.automationintesting.model;

import com.fasterxml.jackson.annotation.JsonProperty;

public class Count {

    @JsonProperty
    private int count;

    public Count(int count) {
        this.count = count;
    }

    public int getCount() {
        return count;
    }
}
