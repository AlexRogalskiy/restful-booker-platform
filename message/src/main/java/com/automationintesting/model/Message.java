package com.automationintesting.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.sql.ResultSet;
import java.sql.SQLException;

public class Message {

    @JsonProperty
    private int messageid;

    @JsonProperty
    private String name;

    @JsonProperty
    private String email;

    @JsonProperty
    private String phone;

    @JsonProperty
    private String subject;

    @JsonProperty
    private String description;

    public Message() {
    }

    public Message(String name, String email, String phone, String subject, String description) {
        this.name = name;
        this.email = email;
        this.phone = phone;
        this.subject = subject;
        this.description = description;
    }

    public Message(ResultSet result) throws SQLException {
        this.messageid = result.getInt("messageid");
        this.name = result.getString("name");
        this.email = result.getString("email");
        this.phone = result.getString("phone");
        this.subject = result.getString("subject");
        this.description = result.getString("description");
    }

    public int getMessageid() {
        return messageid;
    }

    public String getName() {
        return name;
    }

    public String getEmail() {
        return email;
    }

    public String getPhone() {
        return phone;
    }

    public String getSubject() {
        return subject;
    }

    public String getDescription() {
        return description;
    }

    public void setMessageid(int messageid) {
        this.messageid = messageid;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public void setPhone(String phone) {
        this.phone = phone;
    }

    public void setSubject(String subject) {
        this.subject = subject;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    @Override
    public String toString() {
        return "Message{" +
                "messageid=" + messageid +
                ", name='" + name + '\'' +
                ", email='" + email + '\'' +
                ", phone='" + phone + '\'' +
                ", subject='" + subject + '\'' +
                ", description='" + description + '\'' +
                '}';
    }
}
