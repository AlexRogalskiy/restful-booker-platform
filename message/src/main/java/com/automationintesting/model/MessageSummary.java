package com.automationintesting.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.sql.ResultSet;
import java.sql.SQLException;

public class MessageSummary {

    @JsonProperty
    private int id;

    @JsonProperty
    private String name;

    @JsonProperty
    private String subject;

    public MessageSummary(ResultSet resultSet) throws SQLException {
        this.id = resultSet.getInt("messageid");
        this.name = resultSet.getString("name");
        this.subject = resultSet.getString("subject");
    }

    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public String getSubject() {
        return subject;
    }

    @Override
    public String toString() {
        return "MessageSummary{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", subject='" + subject + '\'' +
                '}';
    }
}
