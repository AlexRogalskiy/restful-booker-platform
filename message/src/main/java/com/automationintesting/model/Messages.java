package com.automationintesting.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.ArrayList;
import java.util.List;

public class Messages {

    @JsonProperty
    private List<MessageSummary> messages;

    public Messages(List<MessageSummary> messages) {
        this.messages = messages;
    }

    public List<MessageSummary> getMessages() {
        return messages;
    }

}
