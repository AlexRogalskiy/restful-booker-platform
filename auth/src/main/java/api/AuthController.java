package api;

import auth.Tokens;
import model.Auth;
import model.Token;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.config.annotation.CorsRegistry;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

@RestController
public class AuthController {

    @Bean
    public WebMvcConfigurer corsConfigurer() {
        return new WebMvcConfigurer() {
            @Override
            public void addCorsMappings(CorsRegistry registry) {
                String originHost = "http://localhost:3003";

                if(System.getenv("cors") != null){
                    originHost = System.getenv("cors");
                }

                registry.addMapping("/*")
                        .allowedOrigins(originHost)
                        .allowCredentials(true);
            }
        };
    }

    @RequestMapping(value = "/login", method = RequestMethod.POST)
    public ResponseEntity<Token> createToken(@RequestBody Auth auth) {
        if(auth.getUsername().equals("admin") && auth.getPassword().equals("password")){
            return ResponseEntity.ok(new Token(Tokens.create()));
        } else {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).body(null);
        }
    }

    @RequestMapping(value = "/validate", method = RequestMethod.POST)
    public ResponseEntity<Token> validateToken(@RequestBody Token token) {
        if(Tokens.verify(token.getToken())){
            return ResponseEntity.status(HttpStatus.OK).build();
        } else {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
        }
    }

    @RequestMapping(value = "/logout", method = RequestMethod.POST)
    public ResponseEntity clearToken(@RequestBody Token token) {
        Tokens.clear(token.getToken());

        return ResponseEntity.status(HttpStatus.OK).build();
    }

}
