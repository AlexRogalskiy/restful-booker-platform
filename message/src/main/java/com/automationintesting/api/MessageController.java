package com.automationintesting.api;

import com.automationintesting.db.MessageDB;
import com.automationintesting.model.Count;
import com.automationintesting.model.Message;
import com.automationintesting.model.Messages;
import com.automationintesting.requests.AuthRequests;
import com.automationintesting.utils.DatabaseScheduler;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.config.annotation.CorsRegistry;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

import javax.validation.Valid;
import java.sql.SQLException;
import java.util.concurrent.TimeUnit;

@RestController
public class MessageController {

    @Autowired
    private MessageDB messageDB;
    private AuthRequests authRequest;

    @Bean
    public WebMvcConfigurer configurer() {
        DatabaseScheduler databaseScheduler = new DatabaseScheduler();
        databaseScheduler.startScheduler(messageDB, TimeUnit.MINUTES);

        return new WebMvcConfigurer() {
            @Override
            public void addCorsMappings(CorsRegistry registry) {
                String originHost = "http://localhost:3003";

                if(System.getenv("cors") != null){
                    originHost = System.getenv("cors");
                }

                registry.addMapping("/*")
                        .allowedMethods("GET", "POST", "DELETE", "PUT")
                        .allowedOrigins(originHost)
                        .allowCredentials(true);
            }
        };
    }

    public MessageController() throws SQLException {
        this.authRequest = new AuthRequests();
    }

    @RequestMapping(value = "/", method = RequestMethod.GET)
    public ResponseEntity<Messages> getMessages() throws SQLException {
        Messages messages = new Messages(messageDB.queryMessages());

        return ResponseEntity.ok(messages);
    }

    @RequestMapping(value = "/count", method = RequestMethod.GET)
    public ResponseEntity<Count> getCount() throws SQLException {
        Count count = new Count(messageDB.getUnreadCount());

        return ResponseEntity.ok(count);
    }

    @RequestMapping(value = "/{id:[0-9]*}", method = RequestMethod.GET)
    public Message getMessage(@PathVariable(value = "id") int id) throws SQLException {
        Message queriedMessage = messageDB.query(id);

        return queriedMessage;
    }

    @RequestMapping(value = "/", method = RequestMethod.POST)
    public ResponseEntity<Message> createMessage(@Valid @RequestBody Message message) throws SQLException {
        Message body = messageDB.create(message);
        return ResponseEntity.ok(body);
    }

    @RequestMapping(value = "/{id:[0-9]*}", method = RequestMethod.DELETE)
    public ResponseEntity deleteMessage(@PathVariable(value = "id") int id, @CookieValue(value ="token", required = false) String token) throws SQLException {
        if(authRequest.postCheckAuth(token)){
            if(messageDB.delete(id)){
                return ResponseEntity.accepted().build();
            } else {
                return ResponseEntity.notFound().build();
            }
        } else {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
        }
    }

    @RequestMapping(value = "/{id:[0-9]*}/read", method = RequestMethod.PUT)
    public ResponseEntity markAsRead(@PathVariable(value = "id") int id, @CookieValue(value ="token", required = false) String token) throws SQLException {
        if(authRequest.postCheckAuth(token)){
            messageDB.markAsRead(id);

            return ResponseEntity.ok().build();
        } else {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
        }
    }
}
