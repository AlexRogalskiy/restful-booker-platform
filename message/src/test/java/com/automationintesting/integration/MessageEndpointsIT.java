package com.automationintesting.integration;

import com.automationintesting.api.MessageApplication;
import com.automationintesting.model.Message;
import com.github.tomakehurst.wiremock.common.ConsoleNotifier;
import com.github.tomakehurst.wiremock.core.WireMockConfiguration;
import com.github.tomakehurst.wiremock.junit.WireMockRule;
import io.restassured.http.ContentType;
import io.restassured.response.Response;
import org.approvaltests.Approvals;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringRunner;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
import static io.restassured.RestAssured.given;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT, classes = MessageApplication.class)
@ActiveProfiles("dev")
public class MessageEndpointsIT {

    @Rule
    public WireMockRule wireMockRule = new WireMockRule(WireMockConfiguration.options().notifier(new ConsoleNotifier(true)).port(3004));

    @Before
    public void setupWiremock(){
        stubFor(post("/auth/validate")
                .withRequestBody(equalToJson("{ \"token\": \"abc123\" }"))
                .willReturn(aResponse().withStatus(200)));
    }

    @Test
    public void createMessage(){
        Message messagePayload = new Message("Mark", "test@email.com", "01234556789", "Subject line goes in here for display", "Description details here to give info on request");

        Response response = given()
                .contentType(ContentType.JSON)
                .body(messagePayload)
                .when()
                .post("http://localhost:3006/message/");

        Approvals.verify(response.getBody().prettyPrint());
    }

    @Test
    public void getMessage(){
        Response response = given()
                .get("http://localhost:3006/message/1");

        Approvals.verify(response.getBody().prettyPrint());
    }

    @Test
    public void getMessages(){
        Response response = given()
                .get("http://localhost:3006/message/");

        Approvals.verify(response.getBody().prettyPrint());
    }

    @Test
    public void getCount(){
        Response response = given()
                .get("http://localhost:3006/message/count");

        Approvals.verify(response.getBody().prettyPrint());
    }

    @Test
    public void deleteMessage(){
        Message messagePayload = new Message("Mark", "test@email.com", "01234556789", "Subject line goes in here for display", "Description details here to give info on request");

        Message createdMessage = given()
                .contentType(ContentType.JSON)
                .body(messagePayload)
                .when()
                .post("http://localhost:3006/message/")
                .getBody().as(Message.class);

        Response response = given()
                .cookie("token", "abc123")
                .delete("http://localhost:3006/message/" + createdMessage.getMessageid());

        assertThat(response.statusCode(), is(202));
    }

    @Test
    public void validationTest(){
        Response response = given()
                .contentType(ContentType.JSON)
                .body("{}")
                .when()
                .post("http://localhost:3006/message/");

        assertThat(response.statusCode(), is(400));
    }

    @Test
    public void markAsReadTest(){
        Message messagePayload = new Message("Mark", "test@email.com", "01234556789", "Subject line goes in here for display", "Description details here to give info on request");

        Message createdMessage = given()
                .contentType(ContentType.JSON)
                .body(messagePayload)
                .when()
                .post("http://localhost:3006/message/")
                .getBody().as(Message.class);

        given()
            .cookie("token", "abc123")
            .put("http://localhost:3006/message/" + createdMessage.getMessageid() + "/read");


        Response response = given()
                .get("http://localhost:3006/message/count");

        Approvals.verify(response.getBody().prettyPrint());
    }

}
