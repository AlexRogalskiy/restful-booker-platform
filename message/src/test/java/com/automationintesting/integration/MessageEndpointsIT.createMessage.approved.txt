{
    "messageid": 2,
    "name": "Mark",
    "email": "test@email.com",
    "phone": "01234556789",
    "subject": "Subject line",
    "description": "Description details here"
}
